### PR TITLE
[cleaner] Skip obfuscation of substrings for hostnames properly

### PR DESCRIPTION
**Pull request by pmoravec**
Leftover from #3403 / #3496 where match_full_words_only is ignored for hostname mapping due to own get_regex_result method.

Resolves: #3593

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [X] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

### CODE REVIEW

File: sos/cleaner/mappings/hostname_map.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def get_regex_result(self, item):
         """
         if '.' in item:
             item = item.replace('.', '(\\.|_)')
-        return re.compile(item, re.I)
+        return super(SoSHostnameMap, self).get_regex_result(item)
```
Comment (jcastill) [inline sos/cleaner/mappings/hostname_map.py:90]: If you want to return that way, perhaps add:

`# pylint: disable=super-with-arguments`

To the end of the line?

Comment (arif-ali) [reply sos/cleaner/mappings/hostname_map.py:90]: or better, and then we don't need to disable it :)

```python
return super().get_regex_result(item)
```

**Review by jcastill — APPROVED** [review]

**Comment by arif-ali** [comment]
I think, that makes sense to me, is it worth updating the tests to reflect this at all?

**Comment by pmoravec** [comment]
> I think, that makes sense to me, is it worth updating the tests to reflect this at all?

I came to the same conclusion - that `someHOST-something`  does *not* contain `HOST` as a *hostname*. It contains it as a substring, but not as hostname.

So:
- the original reproducer https://github.com/sosreport/sos/pull/3593#issuecomment-2045819623 shows a valid problem where hostname `port` should not be obfuscated from `sosreport-host0` string.
- but the current patch is wrong, as it does not obfuscate `cleanertest` hostname from `lvm_cleanertest_2015` string

So I will come up with a better patch.

**Comment by pmoravec** [comment]
This is tricky :) Let see what tests will return, but `item = rf'(?=\b|_|-){re.escape(item)}(?=\b|_|-)'` is simply wrong. We test positive lookahead - twice! - but dont test lookbehind. That makes a lot of confusion where `my-hostname_TEST` finds `hostname`, but `my_hostname_TEST` doesnt find it.

OK, let try positive lookbehind `(?<=\b|_|-)` - but Python denies it as "look-behind requires fixed-width pattern".

Let re-formulate it to negative lookbehind :) `(?<![A-Za-z0-9])` . Since alphanumeric characters (*not* `\w`, that's another trick) are exactly those forbidden before a full word match.

So additionally to the patch, we must:

```
diff --git a/sos/cleaner/mappings/__init__.py b/sos/cleaner/mappings/__init__.py
--- a/sos/cleaner/mappings/__init__.py
+++ b/sos/cleaner/mappings/__init__.py
@@ -99,7 +99,7 @@ class SoSMap():
         :rtype:         ``re.Pattern``
         """
         if self.match_full_words_only:
-            item = rf'(?=\b|_|-){re.escape(item)}(?=\b|_|-)'
+            item = rf'(?!<[A-Za-z0-9]){re.escape(item)}(?=\b|_|-)'
         else:
             item = re.escape(item)
         return re.compile(item, re.I)
```

**Comment by pmoravec** [comment]
FYI I made a test script similar to https://github.com/sosreport/sos/pull/3766#issuecomment-2370595072 : 

```
import re

secret = "SECRET"

REG = rf'(?<![A-Za-z0-9]){re.escape(secret)}(?=\b|_|-)'

test_strings = (
        (True, 'test SECRET:something'),
        (True, 'test:SECRET something'),
        (True, 'test my_SECRET_TEST will be cleaned?'),
        (True, 'test my-SECRET_TEST will be cleaned?'),
        (False, 'test mySECRET-should stay'),
        (True, 'test SECRET-cleaned'),
        (True, 'SECRET-cleaned also'),
        (False, 'SECRETNOT-cleaned'),
        (True, 'SECRET_cleaned underscore'),
        (True, 'SECRET.cleaned as well'),
        (True, 'cleaned.SECRET also'),
        (False, 'mySECRET.notcleaned hostname'),
)

intro=f"Matching against {REG}"
print(intro)
print("-"*len(intro))
print()

for match, line in test_strings:
    search = re.search(REG, line)
    print(f"{line}\t:\t\t{search}")
    if (None!=search) != match:
        print(f"\t\tERROR {line} should {'' if match else 'not '}match, but re.search={search}")
```

The script tests if given word (`SECRET`) is matched as a full word in several lines by proper R.E..

**Comment by arif-ali** [comment]
I've put this through regex101 (always my fallback these days) too ;), and works as expected

**Comment by pmoravec** [comment]
Once I fixed the stupid typo, it should pass tests now :)

**Comment by pmoravec** [comment]
The failing test is a false alarm: `sosreport-host0-2024-11-27-pyqgxyc/sos_commands/systemd/systemctl_status_--all` file contains log:

```
Nov 27 12:49:10 cirrus-task-5243713155235840Unit display-manager.service could not be found.
```

and the test finds the hostname `cirrus-task-5243713155235840` there. But hostname is full word matching.

OK, next round of PR will fix the test :)

**Review by arif-ali — APPROVED** [review]

**Comment by pmoravec** [comment]
Since I think I touch a fragile&core part of cleaner (how explicitly we search for matches), and the PR fixes three bugs altogether, I would prefer yet another ACK from either @jcastill or @TurboTurtle .

(three bugs: each commit fixes one)

**Comment by TurboTurtle** [comment]
I'll give this a run through this Saturday.

**Comment by jcastill** [comment]
@pmoravec   I ran some very basic tests and the commits seem to work pretty well. I'll see if I have some time this weekend to run something more complex

**Review by TurboTurtle — APPROVED** [review]
LGTM

**Pull request merged**